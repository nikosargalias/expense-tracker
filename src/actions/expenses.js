import generateNewId from "../helpers/generateId";
import store from "../store/configureStore";

// action generator
const addExpense = (expense) => {
  const prevState = store.getState();
  return {
    type: "ADD_EXPENSE",
    expense: {
      description: "",
      note: "",
      amount: 0,
      createdAt: new Date().getTime(),
      ...expense,
      id: generateNewId(prevState.expenses),
    },
  };
};

const removeExpense = (id) => {
  return {
    type: "REMOVE_EXPENSE",
    id,
  };
};

const editExpense = ({ id, updates } = {}) => {
  console.log(id, updates);
  return {
    type: "EDIT_EXPENSE",
    id,
    updates,
  };
};

export { addExpense, removeExpense, editExpense };
