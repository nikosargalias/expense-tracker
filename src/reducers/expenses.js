const expensesReducerDefaultState = [];

const expenseReducer = (
  prevState = expensesReducerDefaultState,
  { type, expense, id, updates }
) => {
  switch (type) {
    case "ADD_EXPENSE":
      return [...prevState, expense];
    case "REMOVE_EXPENSE":
      return prevState.filter((expense) => expense.id !== id);
    case "EDIT_EXPENSE":
      return prevState.map((expense) => {
        console.log(id);
        if (expense.id === id) {
          return {
            ...expense,
            ...updates,
          };
        }
        return expense;
      });
    default:
      return prevState;
  }
};

export { expenseReducer as default };
