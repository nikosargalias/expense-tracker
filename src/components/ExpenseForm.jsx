import "react-dates/initialize";
import React from "react";
import {
  DateRangePicker,
  SingleDatePicker,
  DayPickerRangeController,
} from "react-dates";

import moment from "moment";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense?.description || "",
      amount: props.expense?.amount || "",
      note: props.expense?.note || "",
      createdAt: props.expense?.createdAt
        ? moment(props.expense.createdAt)
        : moment(),
      submitError: false,
      calendarFfocused: false,
    };
  }
  onChangeDescription(e) {
    const description = e.target.value;
    this.setState(() => ({ description: description }));
  }
  onChangeAmount(e) {
    const input = e.target.value;

    if (!input.toString().match(/^[0-9]*(\.\d{0,2})?$/)) {
      console.error(
        "invalid number, must contain a maximum of two decimal places"
      );
      return;
    }

    let amount = parseFloat(input) || "";
    this.setState(() => ({ amount: amount }));
  }
  onChangeNote(e) {
    const note = e.target.value;
    this.setState(() => ({ note: note }));
  }

  onSubmitExpense(e) {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ submitError: true }));
      return;
    }
    this.props.onSubmit({
      description: this.state.description,
      note: this.state.note,
      amount: this.state.amount,
      createdAt: this.state.createdAt.valueOf(),
    });
  }

  onDateChange(createdAt) {
    if (createdAt) {
      this.setState({ createdAt });
    }
  }
  onFocusChange = ({ focused }) => this.setState({ calendarFocused: focused });
  render() {
    return (
      <form className="form" onSubmit={this.onSubmitExpense.bind(this)}>
        <div className="form__group">
          <label htmlFor="description">Description</label>
          <input
            onChange={this.onChangeDescription.bind(this)}
            type="text"
            id="description"
            value={this.state.description}
          />
        </div>
        <div className="form__group">
          <label htmlFor="amount">Amount</label>
          <input
            onChange={this.onChangeAmount.bind(this)}
            type="number"
            id="amount"
            value={this.state.amount}
          />
        </div>
        <div className="form__group">
          <label htmlFor="note">Note</label>
          <textarea
            onChange={this.onChangeNote.bind(this)}
            placeholder="Additional notes about expense (optional)"
            id="note"
            value={this.state.note}
          />
        </div>
        <div className="form__group">
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange.bind(this)}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange.bind(this)}
            id="your_unique_id"
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
        </div>
        {this.state.submitError && (
          <p className="form__error">Please fill in all fields</p>
        )}
        <button type="submit">
          {this.props.expense ? "Edit Expense" : "Add Expense"}
        </button>
      </form>
    );
  }
}

export default connect()(ExpenseForm);
