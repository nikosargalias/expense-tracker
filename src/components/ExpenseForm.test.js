import React from "react";
import { shallow } from "enzyme";
import { ExpenseForm } from "./ExpenseForm";
import expenses from "../fixtures/expenses";

test("should render ExpenseForm correctly", () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test("should render expense form with expenses[1] data", () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render expense form with expenses[2] data", () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[2]} />);
  expect(wrapper).toMatchSnapshot();
});
