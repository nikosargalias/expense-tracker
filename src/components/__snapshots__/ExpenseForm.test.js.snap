// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__group"
  >
    <label
      htmlFor="description"
    >
      Description
    </label>
    <input
      id="description"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form__group"
  >
    <label
      htmlFor="amount"
    >
      Amount
    </label>
    <input
      id="amount"
      onChange={[Function]}
      type="number"
      value=""
    />
  </div>
  <div
    className="form__group"
  >
    <label
      htmlFor="note"
    >
      Note
    </label>
    <textarea
      id="note"
      onChange={[Function]}
      placeholder="Additional notes about expense (optional)"
      value=""
    />
  </div>
  <div
    className="form__group"
  >
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="your_unique_id"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <button
    type="submit"
  >
    Add Expense
  </button>
</form>
`;

exports[`should render expense form with expenses[1] data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__group"
  >
    <label
      htmlFor="description"
    >
      Description
    </label>
    <input
      id="description"
      onChange={[Function]}
      type="text"
      value="drink"
    />
  </div>
  <div
    className="form__group"
  >
    <label
      htmlFor="amount"
    >
      Amount
    </label>
    <input
      id="amount"
      onChange={[Function]}
      type="number"
      value={100}
    />
  </div>
  <div
    className="form__group"
  >
    <label
      htmlFor="note"
    >
      Note
    </label>
    <textarea
      id="note"
      onChange={[Function]}
      placeholder="Additional notes about expense (optional)"
      value="nothing"
    />
  </div>
  <div
    className="form__group"
  >
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1969-12-29T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="your_unique_id"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <button
    type="submit"
  >
    Edit Expense
  </button>
</form>
`;

exports[`should render expense form with expenses[2] data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__group"
  >
    <label
      htmlFor="description"
    >
      Description
    </label>
    <input
      id="description"
      onChange={[Function]}
      type="text"
      value="Rent"
    />
  </div>
  <div
    className="form__group"
  >
    <label
      htmlFor="amount"
    >
      Amount
    </label>
    <input
      id="amount"
      onChange={[Function]}
      type="number"
      value={4500}
    />
  </div>
  <div
    className="form__group"
  >
    <label
      htmlFor="note"
    >
      Note
    </label>
    <textarea
      id="note"
      onChange={[Function]}
      placeholder="Additional notes about expense (optional)"
      value="Expensive"
    />
  </div>
  <div
    className="form__group"
  >
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-03T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="your_unique_id"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <button
    type="submit"
  >
    Edit Expense
  </button>
</form>
`;
